First version of the property editor

zoltanszocs opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and
review the checklist below.

Contributors guide: ./CONTRIBUTING.md
-->

## Checklist
<!--
Remove items that do not apply. For completed items, change [ ] to [x] or you can click the checkboxes once your 
pull-request is published.
-->

- [ ] Keep pull requests small so they can be easily reviewed.
- [ ] Update the documentation, please follow the [numpy style guide](https://numpydoc.readthedocs.io/en/latest/format.html).
- [ ] Update tests.
- [ ] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `feature`, `ui`, `change`, `documentation`, `breaking`, `ci`
      as they show up in the changelog.
- [ ] Link this PR to related issues by using [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
- [ ] Add a milestone to the PR for the intended release.
- [ ] Request a review from another developer.

<!--
NOTE: these things are not required to open a PR and can be done afterwards,
while the PR is open.
-->

zoltanszocs committed:
First version of the property editor

cmutel commented on activity_browser/layouts/tabs/activity.py:
```diff
@@ -39,7 +46,7 @@ def open_activity_tab(self, key: tuple, read_only: bool = True) -> None:
             act = bd.get_activity(key)
-            if not bc.is_technosphere_activity(act):
+            if act.get("type") not in bd.labels.node_types:
                 return
```
Can we please give some feedback here instead of doing nothing? It is currently very confusing. For example, a popup saying that the current node can't be displayed because it isn't a process or a product.

cmutel commented:
This is a great start!

<img width="848" alt="Screenshot 2024-08-16 at 14 22 53" src="https://github.com/user-attachments/assets/6eca351e-b2a1-410f-8c7f-d09e744a1b37">

- [x] The "properties" button text is very small - this needs to be bigger, and should be "Process Properties". 
- [x] The button should also have an outline to make it clear that it is a clickable button.
- [x] We should be able to right click on a row in the "Products" table and get a similar properties table for that specific product (you can use `bw2data.get_node()` to get the product object and retrieve its properties). I don't think we need to cache the properties for each row, we can generate this on demand by querying the database.

zoltanszocs committed:
Fix log file encoding

zoltanszocs committed:
No editing for read only, change save button caption

zoltanszocs committed:
Hanlde modified and duplicate entries

zoltanszocs committed:
Add delete button and reworked colors

cmutel commented:
@zoltanszocs Looking good, and the functionality works as expected. Please see my [earlier comments](https://github.com/cauldron/activity-browser/pull/42#issuecomment-2293421571) on the next set of things to work on. Thanks!

zoltanszocs committed:
Correct setData to use beginInsertRows

zoltanszocs committed:
Move the model-table-delegate code to its place

zoltanszocs committed:
Add message box when activity can not be opened

zoltanszocs committed:
Change properties to QPushbutton from toolbar button

zoltanszocs committed:
Add status label to property dialog

zoltanszocs committed:
No delete buttons in read-only

zoltanszocs committed:
No delete button for the last row

zoltanszocs committed:
Fixes for empty key rows

Show empty key rows like deleted ones
Do not toggle the delete flag on empty key rows
Make last row last column not editable (because editing it shows then hides the delete button)

zoltanszocs committed:
Move the property editor to ui.widgets

zoltanszocs committed:
Add context menu entry to edit properties of products

zoltanszocs committed:
Change delete button icon to the delete icon already used in AB

cmutel commented:
- [x] @cmutel need to fix https://github.com/brightway-lca/brightway2-data/issues/182

cmutel committed:
Add `Functional` column for products

cmutel commented:
Great; let's add a bit and then we can merge this and do the next stage.

<img width="1882" alt="Screenshot 2024-08-21 at 09 19 37" src="https://github.com/user-attachments/assets/c419b011-02dc-4a66-bb04-1aa2472e13e9">

- [x] 1. Please move the "Properties" button from the right to the left, right after the graph traversal button.
- [x] 2. I added a column "Functional" to the products table in https://github.com/cauldron/activity-browser/pull/42/commits/59eddc29946fb1e99cc1b3edf8b009255f52667b; this needs to be improved by moving from `StringDelegate` to `BooleanDelegate` (not yet defined), which should display with some subtle color coding (i.e. a little bit of green or red).
- [x] 3. @cmutel Add a column for default allocation factor. Blocked currently by https://github.com/brightway-lca/multifunctional/issues/19.

zoltanszocs committed:
Change postion of properties button

zoltanszocs committed:
Add spacing to the activity toolbar

cmutel committed:
Add allocation factor table

zoltanszocs committed:
Fix stylesheet

zoltanszocs committed:
Add functional column

zoltanszocs committed:
Merge branch 'zoltan/multifunctional-properties' of https://github.com/zoltanszocs/activity-browser into zoltan/multifunctional-properties

cmutel committed:
Add `functional` as mofidiable exchange attr

zoltanszocs committed:
Merge branch 'zoltan/multifunctional-properties' of https://github.com/zoltanszocs/activity-browser into zoltan/multifunctional-properties

zoltanszocs committed:
Make functional column editable

zoltanszocs committed:
Make column 3 editable using a checkbox in Product model

zoltanszocs committed:
Move the read only flag to the base model

This allows to prevent edits at the model level

cmutel commented on activity_browser/actions/exchange/exchange_properties.py:
```diff
@@ -0,0 +1,26 @@
+from activity_browser.ui.widgets.property_editor import PropertyEditor
+
+
+class ExchangeProperties(ABAction):
```
Sorry, I can't understand this code on its own. An `Exchange` is very different from an `Activity`, and both the labels and the code itself seem to mix the two. Type hinting isn't really helping here either :)

zoltanszocs replied:
Hmm.. I seem to have a misconception about Exchanges and Activities. From the code I thought activities are subtypes of exchanges. I will reevaluate the code.

cmutel commented on activity_browser/layouts/tabs/activity.py:
```diff
@@ -349,3 +343,9 @@ def save_splitter_state(self):
         ).hex()
         ab_settings.write_settings()
+
+    def open_properties(self):
```
This is a really weird naming convention - it at least needs to be commented on why this is called "open_..."

zoltanszocs replied:
open_property_editor ?

cmutel replied:
Sure, that works.

cmutel commented on activity_browser/ui/tables/activity.py:
```diff
@@ -91,7 +98,28 @@ def get_interpreter(self):
+        else:
+            return [self.model.get_exchange(self.currentIndex())]
+    
+    def set_read_only(self, read_only: bool):
```
This name needs to be easier to understand - maybe `set_read_only_flag` or `set_read_only_state`?

zoltanszocs replied:
set_read_only_flag it is

cmutel commented on activity_browser/ui/tables/activity.py:
```diff
@@ -102,7 +130,9 @@ def __init__(self, parent=None):
         self.setItemDelegateForColumn(1, StringDelegate(self))
         self.setItemDelegateForColumn(2, StringDelegate(self))
-        self.setItemDelegateForColumn(3, FormulaDelegate(self))
+        self.setItemDelegateForColumn(3, StringDelegate(self))
```
Shouldn't this be a new `BooleanDelegate`? Or am I missing something somewhere else? Sorry if so...

zoltanszocs replied:
Good catch! I just saw this today. In fact as I used the built-in checkbox delegate, no delegate needs to be set here.

cmutel replied:
Even better :)

cmutel commented:
We need one more big step here: If I understand correctly, we have a property editor for the current process or product being viewed, and the ability to edit the properties of the products or processes being linked to. However, we also need to be able to edit the *edges* (called `Exchanges` in the codebase). So this menu:

<img width="816" alt="Screenshot 2024-08-22 at 13 34 45" src="https://github.com/user-attachments/assets/76c4c78e-13ec-42e6-99b6-cc62752fd846">

Needs to have two options: "Node properties" and "Edge properties", with the code saving to either the appropriate `Node` object, or to the `Edge`.

cmutel committed:
Squashed commit of the following:

commit 29f7a87faf0d86435a42b945b918bfce5acf7853
Merge: c2846c3b 7a5ddbfa
Author: Chris Mutel <cmutel@gmail.com>
Date:   Thu Aug 22 13:40:32 2024 +0200

    Merge pull request #44 from zoltanszocs/zoltan/default-allocation

    Add default allocation column to databases

commit 7a5ddbfa063cdb221a5d2a76ffdb0f9f1d4ca521
Author: Chris Mutel <cmutel@gmail.com>
Date:   Thu Aug 22 13:40:05 2024 +0200

    Change default allocation value

commit aa3a8da1bd5b9c9fc51a34ec8d697e9d862f1e84
Author: Zoltan Szocs <zoltan.szocs@gmail.com>
Date:   Thu Aug 22 14:08:33 2024 +0300

    Add default allocation column to databases

zoltanszocs committed:
Renames, minor changes after review

zoltanszocs committed:
Add separate property editors for nodes and edges
